File: src/main/java/com/locusdev/hlatool/Allele.java
package com.locusdev.hlatool;

import java.util.List;

/**
 * User: alexfurman
 * Date: Apr 21, 2010
 * Time: 2:32:05 PM
 */
public class Allele {
    private String name;
    private String sequence;

    public Allele(String name, String sequence) {
        this.name = name;
        this.sequence = sequence;
    }

    public String getName() {
        return name;
    }

    public String getSequence() {
        return sequence;
    }

    public boolean matchesMutation(JMutation mutation) {
        return mutation.matches(getSequence());
    }

    public boolean baseSequenced(Integer input) {
        if (getSequence().length() < (input - 1)) {
            throw new IllegalArgumentException(this + " sequence out of bounds: " + input + ", length: " +
                    getSequence().length());
        }
        return getNucleotide(input) != '*';
    }

    public char getNucleotide(Integer input) {
        return getSequence().charAt(input);
    }

    public boolean matchesMutations(List<JMutation> mutations) {
        for (JMutation mutation : mutations) {
            if (!matchesMutation(mutation)) {
                return false;
            }
        }
        return true;
    }

    @Override
    public String toString() {
        return "Allele{" +
                "name='" + name + '\'' +
                '}';
    }
}

File: src/main/scala/com/locusdev/hlatool/Allele.scala
package com.locusdev.hlatool

/**
 * Created by IntelliJ IDEA.
 * User: alexfurman
 * Date: May 18, 2010
 * Time: 1:28:35 PM
 * To change this template use File | Settings | File Templates.
 */

case class Allele(name : String, blocks : scala.collection.Map[String, String]) {

  def sequence = Haplotyper.combineBlocks(blocks)

  def baseSequenced(index : Int) = {
    if (sequence.length < index - 1) throw new IllegalArgumentException(this + " sequence out of bounds: " + index + ", length: " + sequence.length)
    sequence.charAt(index) != '*'
  }

  def matches(mutations : List[Mutation]) : Boolean = {
    !(mutations exists { mutation => !matches(mutation) })
  }

  

  def matches(mutation : Mutation) : Boolean = (sequence.charAt(mutation.index) == mutation.nucleotide)
}